File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/domain/interactor/location/GetPlaceBufferUseCase.java
package com.lattis.ellipse.domain.interactor.location;

import com.google.android.libraries.places.api.model.Place;
import com.lattis.ellipse.domain.executor.PostExecutionThread;
import com.lattis.ellipse.domain.executor.ThreadExecutor;
import com.lattis.ellipse.domain.interactor.UseCase;
import com.lattis.ellipse.domain.repository.LocationRepository;

import java.util.ArrayList;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;

public class GetPlaceBufferUseCase extends UseCase<Place> {

    private LocationRepository locationRepository;
    private String placeId;

    @Inject
    protected GetPlaceBufferUseCase(ThreadExecutor threadExecutor,
                               PostExecutionThread postExecutionThread,
                               LocationRepository locationRepository) {
        super(threadExecutor, postExecutionThread);
        this.locationRepository = locationRepository;
    }

    public GetPlaceBufferUseCase withPlaceId(String placeId){
        this.placeId = placeId;
        return this;
    }

    @Override
    protected Observable<Place> buildUseCaseObservable() {
        return this.locationRepository.getPlaceBuffer(placeId);
    }

    @Override
    public Disposable execute(DisposableObserver<Place> UseCaseSubscriber) {
        return this.buildUseCaseObservable()
                .subscribeWith(UseCaseSubscriber);
    }


}

File: velolabs_repos/lattis_android_v2/domain/src/main/java/com/lattis/domain/usecase/location/GetPlaceBufferUseCase.kt
package com.lattis.domain.usecase.location

import com.google.android.libraries.places.api.model.Place
import com.lattis.domain.executor.PostExecutionThread
import com.lattis.domain.executor.ThreadExecutor
import com.lattis.domain.repository.LocationRepository
import com.lattis.domain.usecase.base.UseCase
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.observers.DisposableObserver
import javax.inject.Inject


class GetPlaceBufferUseCase @Inject
constructor(threadExecutor: ThreadExecutor,
            postExecutionThread: PostExecutionThread,
            private val locationRepository: LocationRepository
) : UseCase<Place>(threadExecutor, postExecutionThread) {
    private lateinit var placeId: String

    fun withPlaceId(placeId: String): GetPlaceBufferUseCase {
        this.placeId = placeId
        return this
    }

    override fun buildUseCaseObservable(): Observable<Place> {
        return this.locationRepository.getPlaceBuffer(placeId)
    }

    override fun execute(UseCaseSubscriber: DisposableObserver<Place>): Disposable {
        return this.buildUseCaseObservable()
                .subscribeWith(UseCaseSubscriber)
    }


}

